cmake/modules/extensions.cmake

```diff
@@ -2487,6 +2487,72 @@ function(zephyr_file_copy oldname newname)
   endif()
 endfunction()
 
+# Usage:
+#   zephyr_file_suffix(<filename>)
+#
+# Zephyr file add suffix extension.
+# This function will check the provied filename or list of filenames to see if they have a
+# `_<FILE_SUFFIX>` extension to them and if so, updates the supplied variable/list with the new
+# path/paths.
+#
+# <filename>: Variable or list of absolute path filename(s) which should be checked and updated if
+#             there is a filename which has the FILE_SUFFIX present.
+#
+# Returns an updated variable/list of absolute path(s)
+#
+function(zephyr_file_suffix filename)
+  if(NOT DEFINED FILE_SUFFIX OR NOT DEFINED ${filename})
+    # If the file suffix variable is not known then there is nothing to do, return early
+    return()
+  endif()
+
+  # Check if the user supplied a list or a single item
+  list(LENGTH ${filename} list_length)
+
+  if(list_length GREATER 1)
+    set(tmp_list ${${filename}})
+    set(tmp_new_list)
+
+    # Loop for each list element
+    while(list_length GREATER 0)
+      list(POP_FRONT tmp_list tmp_element)
+      zephyr_file_suffix(tmp_element)
+      list(APPEND tmp_new_list ${tmp_element})
+      math(EXPR list_length "${list_length} - 1")
+    endwhile()
+
+    # Update supplied variable if it differs
+    if(NOT "${${filename}}" STREQUAL "${tmp_new_list}")
+      set(${filename} "${tmp_new_list}" PARENT_SCOPE)
+    endif()
+  else()
+    if("${${filename}}" STREQUAL "")
+      # Skip checking empty variables
+      return()
+    endif()
+
+    # Search for the full stop so we know where to add the file suffix before the file extension
+    string(FIND "${${filename}}" "." full_stop_position REVERSE)
+    if("${full_stop_position}" EQUAL "-1")
+      set(new_filename "${${filename}}_${FILE_SUFFIX}")
+    else()
+      string(SUBSTRING "${${filename}}" 0 ${full_stop_position} filename_beginning)
+      string(SUBSTRING "${${filename}}" ${full_stop_position} -1 filename_ending)
+      set(new_filename "${filename_beginning}_${FILE_SUFFIX}${filename_ending}")
+    endif()
+
+    # Check if the provided path is relative, if so, prepend the application configuration
+    # directory to it
+    cmake_path(IS_RELATIVE new_filename relative_path)
```

Comment:
also, let's not re-do old errors.

We have had lengthy discussion as to `relative to what` means, and therefore we only allow relative paths for user defined settings.
All settings done within CMake, like a `set(file path)` should be done using abs path, for example `set(file ${CMAKE_CURRENT_SOURCE_DIR}/path)`. No relative paths in such cases.

Take a look here: https://github.com/zephyrproject-rtos/zephyr/blob/62809d758160baed837b6589ff8122cedc29951d/cmake/modules/extensions.cmake#L2411-L2412

https://github.com/zephyrproject-rtos/zephyr/blob/62809d758160baed837b6589ff8122cedc29951d/cmake/modules/extensions.cmake#L2482-L2485

Reply:
This is needed because this runs before the path has been transformed by the other functions, so it need to check if a filename with a suffix exists before or it will miss. It's been months since I looked but it was along the lines of if you have a `prj_<suffix>.conf` in a sysbuild configuration folder but no `prj.conf` then this is needed because without it, sysbuild will look for a `prj.conf` file which fails then revert back to the normal application's configuration directory which then won't find the out-of-tree `prj_<suffix>.conf` file

Reply:
Here we go, creating `prj2.conf` in `samples/subsys/mgmt/mcumgr/smp_svr` then running using `cmake -GNinja -DBOARD=nrf52840dk_nrf52840 -DAPP_DIR=.. -DFILE_SUFFIX=test -Dsmp_svr_CONF_FILE="prj2.conf" /tmp/aa/zephyr/share/sysbuild/`:
```
   *****************************
   * Running CMake for smp_svr *
   *****************************
...
CMake Warning at /tmp/aa/zephyr/cmake/modules/extensions.cmake:2629 (message):
  Check: prj2.conf
...
CMake Warning at /tmp/aa/zephyr/cmake/modules/extensions.cmake:2644 (message):
  RELATIVE SKIP
...
Loaded configuration '/tmp/aa/zephyr/boards/arm/nrf52840dk_nrf52840/nrf52840dk_nrf52840_defconfig'
Merged configuration '/tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/prj2.conf'
Merged configuration '/tmp/aa/zephyr/samples/subsys/mgmt/mcumgr/smp_svr/_AA/smp_svr/zephyr/.config.sysbuild'
```

Reply:
> `-DFILE_SUFFIX=test -Dsmp_svr_CONF_FILE="prj2.conf"`

This is a mess imo, if an explicit conf file has been set with `CONF_FILE` by the user, then we should not start messing with that.
In this specific case, the user should use `-Dsmp_svr_CONF_FILE="prj2_test.conf"`

`FILE_SUFFIX` should be for the cases where the system is looking up a default set of files because none is given by user, meaning for example system sould be using `prj.conf`, `boards/<board>.overlay`, etc., but because user has set `FILE_SUFFIX=foo`, then `prj_foo.conf`, `boards/<board>_foo.overlay`, etc is looked up in addition.

Reply:
> meaning for example system sould be using prj.conf, boards/<board>.overlay, etc., but because user has set FILE_SUFFIX=foo, then prj_foo.conf, boards/<board>_foo.overlay, etc is looked up in addition

Do not agree, this is a worthless feature if it can only add files **in addition**, the point would be to allow for different configuration regardless without imposing anything on the user, with this rule it means they must construct their project in a way that essentially prj.conf is useless and a FILE_SUFFIX must be provided for every build

Reply:
> an example, a user wants to build with debug confs, except for foo image which should still use prj.conf.
> Such user should be able to do:

Well no a user should target it properly using `cmake -GNinja ... -Dabc_FILE_SUFFIX=debug <blah>` or not have a `prj_debug.conf` for `foo`

Reply:
> Well no a user should target it properly using `cmake -GNinja ... -Dabc_FILE_SUFFIX=debug <blah>` or not have a `prj_debug.conf` for `foo`

There can be many reason why you want to do a given build, let it be `debug`, `release`, `fast`, etc, and have everything constant but point a single image to be build different.
Like:
```
west build --sysbuild -DSUFFIX=foo ....
```
Keep everything the same, but now adjust only for bar image:
```
west build --sysbuild -DSUFFIX=foo -Dbar_<setting>=<test-file> ....
```

The more areas that are making use of file suffix, the higher the chance that a user needs just to explicitly point to a given file grows.

Therefore, if a user explicit points to a given file, then we should honor that.

> with this rule it means they must construct their project in a way that essentially prj.conf is useless and a FILE_SUFFIX must be provided for every build.

why does a user need to construct their project in a more special way than before ?
System still fall-back to default files, for example `prj.conf` if `prj_<suffix>.conf` is not found.

Only the case where user explicit specify a given file using `-D<setting>=<use-this-file>`, then we should honor that file.

Reply:
Because if you want to build normally, you do that, if you want to build in debug mode and files are set in e.g. cmake or a module for what to include, you then need to fork those modules and add code for handling it. Let's say someone creates a module `logger` and has `release` and `debug` build for it whereby `release` disables it and `debug` enables it and the module just adds the base .conf file, with my way `-DFILE_SUFFIX=release` and `-DFILE_SUFFIX=debug` work out of the box, with your way it cannot work at all